pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -594,57 +846,144 @@ func (vs *VSphere) DiskIsAttached(volPath string, nodeName k8stypes.NodeName) (b
 	}
 	requestTime := time.Now()
 	isAttached, err := diskIsAttachedInternal(volPath, nodeName)
+	if err != nil {
+		isManagedObjectNotFoundError, err := vs.retry(nodeName, err)
+		if isManagedObjectNotFoundError {
+			if err == vclib.ErrNoVMFound {
+				isAttached, err = false, nil
+			} else if err == nil {
+				isAttached, err = diskIsAttachedInternal(volPath, nodeName)
+			}
+		}
+	}
 	vclib.RecordvSphereMetric(vclib.OperationDiskIsAttached, requestTime, err)
 	return isAttached, err
 }
 
 // DisksAreAttached returns if disks are attached to the VM using controllers supported by the plugin.
+// 1. Converts volPaths into canonical form so that it can be compared with the VM device path.
+// 2. Segregates nodes by vCenter and Datacenter they are present in. This reduces calls to VC.
+// 3. Creates go routines per VC-DC to find whether disks are attached to the nodes.
+// 4. If the some of the VMs are not found or migrated then they are added to a list.
+// 5. After successful execution of goroutines,
+// 5a. If there are any VMs which needs to be retried, they are rediscovered and the whole operation is initiated again for only rediscovered VMs.
+// 5b. If VMs are removed from vSphere inventory they are ignored.
 func (vs *VSphere) DisksAreAttached(nodeVolumes map[k8stypes.NodeName][]string) (map[k8stypes.NodeName]map[string]bool, error) {
 	disksAreAttachedInternal := func(nodeVolumes map[k8stypes.NodeName][]string) (map[k8stypes.NodeName]map[string]bool, error) {
-		attached := make(map[k8stypes.NodeName]map[string]bool)
-		if len(nodeVolumes) == 0 {
-			return attached, nil
+
```

Comment:
do you have some estimate how much overhead caused by DisksAreAttached? Does it increase with the number of nodes, or disks?

Reply:
Hi Jing,

New logic should be the same if not faster.
According to new logic (NodeManager), VCP now caches VM ManagedObjects in the NodeManager cache. If for some reason VM Managed Objects is not valid anymore (VM migrated to another Datacenters /  vCenter) then only finder will be used; else cached VM managed object will be used.

Earlier (1.8.x code base) everytime finder was used. Time taken by finder was linearly increasing with more number VMs in the vCenter.


Thus, bottom line is, unless VM migrates to another vSphere entity (ESXi cluster, Datacenter, vCeneter) new logic is faster than old logic. If VM migrates time taken is comparable to old logic.